### DISCUSSION TITLE
Maintaining distance between two non-integrated rigid bodies with barostat

### DISCUSSION THREAD

**Discussion by arnau-jr** [glotzerlab/hoomd-blue#1826] [Support] [answered]
I'm building a system comprised of a non-cubic box such that x=y<z with two 'slabs' of gold atoms that cover the entire xy box in such a way that they tile with the xy replicas and cover the entire plane (they are, effectively, walls). The slabs are placed at the two ends of the box such that there is some spacing between the 'outer' sides of the slabs. The slabs are functionalized with ligands in the 'inner' sides and there is water between the functionalized sides as well (not shown in picture)

![image](https://github.com/glotzerlab/hoomd-blue/assets/48213666/f261c228-98ef-44ef-9cf3-c66ac53137de)

The image has two x coordinate replicas to show the tiling.

I set the slabs so they are rigid bodies and since I don't want collisions with the xy replicas or for the outer spacing to change I have decided to exclude the slabs from integration and just treat them as stationary walls.

This is where I'm facing issues. I want to run a NPT simulation in such a way that the outer spacing does not change. By setting `box_dof = [False,False,True,False,False,False]` I can make it so only the z coordinate is rescaled. But since the particle rescaling only applies to the integration group then eventually the rest of the system overlaps with the slabs. If instead I enable rescaling for all particles by doing `rescale_all = True` then I don't get overlaps but the outer spacing between slabs is not constant (due to the rescaling).

Is there such a way to set up this simulation so it works without integrating the slabs themselves? The alternative would be to integrate the slabs and then set up a `Distance` constrain between the centers of the slabs and allow the slabs to horizontally collide with their xy replicas, but this is not ideal.

Thank you!

**Comment by joaander** [comment]
Use `constrain.Rigid` (rigid bodies) for small collections of particles that should move together with both transnational and rotational degrees of freedom. Omit particles from the integration group to grant them no degrees of freedom. Choose one of these two options, not both.

I don't understand how you can both fix the wall spacing AND the pressure. Doesn't fixing the wall spacing imply a constant volume by definition?

The `ContsantPressure` integrator and the `ThermodynamicQuantities` compute in HOOMD-blue has no knowledge of the boundary conditions that you impose. They assumes that the volume of the periodic simulation box is the appropriate volume to use when computing the system pressure: https://hoomd-blue.readthedocs.io/en/v4.7.0/module-md-compute.html#hoomd.md.compute.ThermodynamicQuantities.pressure.

**Reply by arnau-jr** [reply]
Is there any particular reason why doing both is a bad idea? Right now I'm defining the rigid body and excluding the center from integration, this way it's easier to manage. Is excluding the center different than excluding the constituent atoms without defining the rigid body?

My intention is to fix the spacing only on one side of the slabs, not on both. Like you say, fixing both would imply fixing the volume. My intention is for the volume outside the slabs to be constant (by fixing the distance, since x and y are constant) but allow for the interior volume to change.

I'm trying to define both slabs as belonging to a single rigid body instead to two separate ones, then the relative geometry of the plates will be constant while allowing the box to change freely.

**Reply by joaander** [reply, marked as answer]
> Is there any particular reason why doing both is a bad idea? Right now I'm defining the rigid body and excluding the center from integration, this way it's easier to manage. Is excluding the center different than excluding the constituent atoms without defining the rigid body?

You will likely be unable to use MPI domain decomposition. See your related question #1815.

> My intention is to fix the spacing only on one side of the slabs, not on both. Like you say, fixing both would imply fixing the volume. My intention is for the volume outside the slabs to be constant (by fixing the distance, since x and y are constant) but allow for the interior volume to change.
> 
> I'm trying to define both slabs as belonging to a single rigid body instead to two separate ones, then the relative geometry of the plates will be constant while allowing the box to change freely.

This is a very specific set of requirements that you can achieve with a custom Updater.

**Reply by arnau-jr** [reply]
> You will likely be unable to use MPI domain decomposition. See your related question https://github.com/glotzerlab/hoomd-blue/discussions/1815.

You are right. I was aware of this limitation but we currently don't need to use MPI for this particular system, but it is a good point.

> This is a very specific set of requirements that you can achieve with a custom Updater.

I will look into it.
